samza-core/src/main/java/org/apache/samza/clustermanager/ResourceRequestState.java

```diff
@@ -264,20 +266,36 @@ private void clearState() {
   }
 
   /**
-   * Retrieves, but does not remove, the first allocated resource on the specified host.
+   * Retrieves an allocated resource on the specified host.
    *
    * @param host  the host for which a resource is needed.
-   * @return      the first {@link SamzaResource} allocated for the specified host or {@code null} if there isn't one.
+   * @return a {@link SamzaResource} allocated for the specified host or {@code null} if there isn't one.
    */
 
-  public SamzaResource peekResource(String host) {
+  public SamzaResource getResource(String host) {
     synchronized (lock) {
-      List<SamzaResource> resourcesOnTheHost = this.allocatedResources.get(host);
+      List<SamzaResource> resourcesOnPreferredHostBuffer = this.allocatedResources.get(host);
 
-      if (resourcesOnTheHost == null || resourcesOnTheHost.isEmpty()) {
-        return null;
+      if (resourcesOnPreferredHostBuffer != null && !resourcesOnPreferredHostBuffer.isEmpty()) {
+        log.info("Returning a buffered resource for {}", host);
+        return resourcesOnPreferredHostBuffer.get(0);
+      } else {
+        log.info("No resources on preferred-host buffer. Scanning ANY_HOST buffer");
+        return getResourceOnAnyHostBuffer(host);
       }
-      return resourcesOnTheHost.get(0);
+    }
+  }
+
+  private SamzaResource getResourceOnAnyHostBuffer(String host) {
+    log.info("Scanning the ANY_HOST buffer for host {}", host);
```

Comment:
Can remove this line (partially redundant) and add the host name to the info log on line 283 (in getResource).

Reply:
Minor: Maybe just inline this method in the caller.